### ISSUE TITLE
working directory not correct on windows

### ISSUE THREAD

**Issue by yyq2013** [pkg/sftp#217]
The working directory now the sftp server gets is like "/D:/temp", but it can't be open on windows if I use "ls" command to list files under it.
The root cause of it is the method of file **request-server.go**, the _path.IsAbs(p)_ should be **filepath.IsAbs(p)**, I tested it works fine. I don't know how to commit the code or create pull request.
Please help to fix.

`// Makes sure we have a clean POSIX (/) absolute path to work with
func cleanPath(p string) string {
	p = filepath.ToSlash(p)
	if !path.IsAbs(p) {
		p = "/" + p
	}
	return path.Clean(p)
}`

**Comment by puellanivis** [comment]
The behavior of the request-server only recently became actually specified. (And not just kind of “hand-waved”). There are a lot of difficulties in supporting an arbitrary filesystem pathing method. The decision was made that while the SFTP standard does not explicitly list any specific pathing method, that the filesystem-abstracting RequestServer should be using a standardized system. The system that was chosen was the URL/POSIX type slash-only system.

I understand that this might cause a change in behavior particularly in Windows. This was an expected and known issue, we worked really hard to come up with a solution that would work cross-platform and particularly on Windows, but in the end, there existed no such solution.

Your particular case may require specific code refactoring, of which no one can really help with without seeing your code. However, the use of `path.Abs` was intentional, and unfortunately will be the intended behavior moving forward.

**Comment by puellanivis** [comment]
TL;DR: I worked with some people fixing another bug to make sure it worked correctly on windows. RequestServer was always intended to work abstractly upon filenames, and at all times was assuming working in POSIX paths. This is now the defined behavior, and some code may require refactoring to account for this.

**Comment by eikenb** [comment]
Note that there is a bug here though. The leading '/' in '/D:/...' is wrong. For some reason I was under the impression that ToSlash() would strip the volume letter off, converting it to a strictly POSIX path. But it doesn't touch the volume. Should we strip those? Based on our decision to use strictly POSIX paths we should. But if we do that could result in Windows servers not being able to serve files from different volumes, or at least making it much harder to do so.

@puellanivis @LordRalex .. thoughts?

**Comment by LordRalex** [comment]
What generally I see is that conversion at it just turns /C:/ to /C/ which designates the drive being accessed.

This is what cygwin does at least, but uses /cygwindrive/ or some other prefix, but then follows that pattern.

The project we have uses the request server as a chroot effectively, and I do not recall that the Windows version had issues with being fed such paths, but I can re-check.

**Comment by eikenb** [comment]
Looking around it does seem that the closest thing to a standard is just making it part of the path, eg. "C:/..." -> "/C/...".

@yyq2013 Can you work with that?

**Comment by yyq2013** [comment]
POSIX path does not work on my windows 10.

**Comment by puellanivis** [comment]
Actually, the file-path URL scheme recommended by Microsoft is `file:///c:/path/to/file`

> POSIX path does not work on my windows 10.

Yes, raw POSIX paths will not work. You will have to transform it into a local Windows path, just like any HTTP server needs to “mangle” paths in order to get them into a valid Windows path.

You should be able to do this with `if path.IsAbs(path) { path = path[1:] }` followed by perhaps a check just to make sure it is a drive-letter access.

But seriously, there are so many ways that Windows paths can behave, that it is simply intractable to make RequestServer accommodate them. If you are just sharing files anyways, you should likely be using Server, rather than the abstracted RequestServer.

**Comment by eikenb** [comment]
@yyq2013 Even using filepath.IsAbs() wouldn't it still be broken as the paths are delimited by POSIX-style slashes (/) vs Windows which uses backslashes (\)?

**Comment by yyq2013** [comment]
@eikenb my mistake words 

> POSIX path does not work on my windows 10.,

 Only this format doesn't work:  /c:/..., but this c:/... works fine.

**Comment by eikenb** [comment]
If normal slashes work in the path then I guess I'm leaning toward just switching to filepath.IsAbs(). It passes the tests and unlike filepath.Abs(), which we discusses in #207, it doesn't actually screw up the path in some cases. It seems unreasonable to stick with path.IsAbs() when switching has no downside and fixes this issue (which wasn't there before).

@puellanivis @LordRalex Objections?

**Comment by LordRalex** [comment]
All IsAbs does on Linux is look for it to start with a /, which then caused issues with the Cyberduck client that opened up the issue before. It doesn't actually do a full check.

To filepath, /. is Abs, along with /a/b/.. based on how I read that file source.

Not sure it's going to solve all of the issues we have going on here.

**Comment by eikenb** [comment]
path.IsAbs() and filepath.IsAbs() on Unix/Linux do pretty much the same thing.

path.IsAbs()
`func IsAbs(path string) bool {
	return len(path) > 0 && path[0] == '/'
}`

filepath.IsAbs()
`func IsAbs(path string) bool {
	return strings.HasPrefix(path, "/")
}`

The Cyberduck issue in #207 wasn't with IsAbs(), that wasn't used before our fix. The "/." problem was fixed by running Clean() after the '/' was prepended to the path.

Changing the current cleanPath() where it uses path.IsAbs() to filepath.IsAbs() shouldn't affect the behavior on Linux at all (from what it is now).

**Comment by LordRalex** [comment]
Huh, that really seems like it's incorrect, but I'll go with it. Then I can't think of anything, but I would like to play with it first, cause I'd hate to cause more issues from all of this.

**Comment by puellanivis** [comment]
However, `filepath.IsAbs` returns true on windows for `\\UNCHOST\path\to\filename`

**Comment by eikenb** [comment]
> However, filepath.IsAbs returns true on windows for \\UNCHOST\path\to\filename

Could you elaborate? I'm not sure what your point is.

**Comment by eikenb** [comment]
Another reason I'm considering it is just due to the fact that we've got a release now and this kind of breaks the semantic versioning contract. It breaks something that once worked.

**Comment by LordRalex** [comment]
He I think means that a path like \\svr-files\somepath is also valid, which makes sense to me.

**Comment by eikenb** [comment]
But leading with 'However' makes it sound like he thinks this is a problem and I'm not sure why. The filepath.ToSlash() would convert it to `//UNCHOST/path/to/filename` which path.IsAbs() would see as an absolute path the same as filepath.IsAbs().

**Comment by yyq2013** [comment]
func cleanPath(p string) string {
	p = filepath.ToSlash(p)
	if !filepath.IsAbs(p) {
		**p = "/" + p**
	}
	return path.Clean(p)
}

On my windows 10, after toSlash, the path is like C:/temp/, use filepath.IsAbs(p) it returns true, so will not add a slash before the path(final is "C:/temp/") which will work on my PC. But if use path.IsAbs(p) it will returns false, and a slash will be insert at first and final is "/C:/temp/" which will not work.

**Comment by eikenb** [comment]
@yyq2013 I understand the issue and your proposed solution. I was just talking it over with the people who were involved in the change (#207) that broke your setup. I'd like something that can work on windows while minimizing what can only be tested on windows, as I don't have a windows system to use for testing.

Thank you for your patience.

**Comment by puellanivis** [comment]
Minor point: she.

After that, I was just bringing up that Windows filepath.IsAbs considers very complex paths. Just throwing wrenches, and letting other people see if they cause problems. ;)

Your analysis seems good though, and using filepath.IsAbs on this matter is probably the best choice due to semantic versioning expectations.

**Comment by eikenb** [comment]
> Minor point: she.
Sorry, I knew that but wasn't paying close enough attention to catch the typo.

Thanks for the help. I think I will go ahead and make this change as I have yet to come up with substantive downside and neither you nor LordRalex have raised one. Also I don't like that I broke something that was working.

Thanks again and thanks @LordRalex as well. I really appreciate your contributions to these issues.
